### PR TITLE
Capstone review

### PR DESCRIPTION
**Pull request by OnlyUnknown**
The user logs in to the website, only by typing the username .

In the navigation panel, the user can see links to:
Motorcycles/doctors/classes/items that you selected as a theme.
"Reserve" form.
"My reservations".
"Add motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).
"Delete motorcycle/doctor/class/item that you selected as a theme" (the link is visible to everybody).

On the main page, the user can see a [list of motorcycles/doctors/classes/items that you selected as a theme].

 When the user selects a specific item, they can see the [details page] with its full description (skip the "Rotate image" button).
In the details page, the user can click the "Reserve" button (in the design you can see the "Configure" button - please replace it with the "Reserve" button).

When the user clicks the "Add item" link in the navigation panel they can see a form for adding a new item.

Make the app responsive, creating both mobile and desktop versions.

When the user clicks the "Delete item" link in the navigation panel they can see a list of all items with title and "Delete" button.
When the user clicks the "Delete" button, the selected item is marked as removed and does not show on the main list anymore.

To reserve an appointment, the user has to select a date and city (username and selected item are autofilled).
Use the design based on the ["Book a vespa test-ride"](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005583) and add all necessary inputs.
The user can also access the "Reserve" page from the navigation panel. In that case only username is autofilled.

When the user clicks the "My reservations" link in the navigation panel they can see a list of their reservations (with information about item name, date and city).

Implement proper user authentication from the front-end to the server.

### CODE REVIEW

File: doctor_appointment_backend/app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    end
+  end
+
+  def show
+    render json: @user
+  end
+
+  def destroy
+    @user&.destroy
+    render json: { message: 'User deleted!' }
```
Comment (mcrd25) [inline doctor_appointment_backend/app/controllers/api/v1/users_controller.rb:24]: Don't forget error handling

File: doctor_appointment_backend/app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+class Api::V1::UsersController < ApplicationController
+  before_action :set_user, only: %i[show destroy]
+
+  def index
+    @users = User.all.order(created_at: :desc)
+    render json: @users
```
Comment (mcrd25) [inline doctor_appointment_backend/app/controllers/api/v1/users_controller.rb:6]: Don't forget error handling

File: doctor_appointment_frontend/src/components/add.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            />
+          </div>
+
+          <div className="relative mb-4 flex text-white items-center justify-center">
+            <input
+              id="speciality"
+              name="speciality"
+              value={speciality}
+              placeholder="Speciality"
+              required
```
Comment (mcrd25) [inline doctor_appointment_frontend/src/components/add.js:114]: These inputs are very similar, maybe they can be abstracted with subcomponent, array, and map [OPTIONAL]

File: doctor_appointment_frontend/src/components/detail.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+              />
+            </svg>
```
Comment (mcrd25) [inline doctor_appointment_frontend/src/components/detail.js:90]: For cleaner code, you should consider storing your svg separately

**Review by mcrd25 — CHANGES_REQUESTED** [review]
Hi Team,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Meet functional requirements 👍 
- Have tests 👍 
- Have good documentation 👍 

## Required Changes ♻️
_Check the comments under the review._
- You are missing some error handling, I would have pointed out some, but please review for those I missed
- You are missing unit tests for your back-end
FRONT-END
- There are some candidates to make some of these html components reusable.
- 
*  I am not seeing tests for your Front-End code - this is something we expect to be implemented for the final evaluation. Here are some resources you can use:
   *  https://reactjs.org/docs/testing.html
   * https://jestjs.io/docs/tutorial-react
   * https://www.freecodecamp.org/news/testing-react-hooks/
- Don't forget that the application needs to look exactly like design with exception of photos, titles and descriptions

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [OPTIONAL] Should always have positive and negative tests to ensure it works the way you expect it. Eg. What do you expect the output to be when it's not valid or when valid etc.
Consider making all your tests more extensive, not just this one. For this simple application, it may not catch a possible crippling bug, but in bigger applications, it definitely can. Consider this for all projects moving forward.
- [OPTIONAL] removing the `test` directory since you are using `rspec` for testing and not native testing.
- [OPTIONAL] Consider using the `timecop` gem to freeze time when running tests. It can help avoid errors: https://github.com/travisjeffery/timecop
- (frontend) see my comments on SVG and sub-components

**Hope all is well, and that my comments were helpful. 😸**

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
_Also, you do not have to implement all requested changes, as this is up to you in the end, but I strongly recommend you implement them._

**Comment by OnlyUnknown** [comment]
> 

Thank you

> Hi Team,
> 
> Good job so far! There are some issues that you still need to work on to go to the next project but you are almost there!
> 
> ## Highlights
> * Meet functional requirements 👍
> * Have tests 👍
> * Have good documentation 👍
> 
> ## Required Changes ♻️
> _Check the comments under the review._
> 
> * You are missing some error handling, I would have pointed out some, but please review for those I missed
> * You are missing unit tests for your back-end
>   FRONT-END
> * There are some candidates to make some of these html components reusable.
> 
> * I am not seeing tests for your Front-End code - this is something we expect to be implemented for the final evaluation. Here are some resources you can use:
>   
>   * https://reactjs.org/docs/testing.html
>   * https://jestjs.io/docs/tutorial-react
>   * https://www.freecodecamp.org/news/testing-react-hooks/
> 
> * Don't forget that the application needs to look exactly like design with exception of photos, titles and descriptions
> 
> ## Optional suggestions
> _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
> 
> You can also consider:
> 
> * [OPTIONAL] Should always have positive and negative tests to ensure it works the way you expect it. Eg. What do you expect the output to be when it's not valid or when valid etc.
>   Consider making all your tests more extensive, not just this one. For this simple application, it may not catch a possible crippling bug, but in bigger applications, it definitely can. Consider this for all projects moving forward.
> * [OPTIONAL] removing the `test` directory since you are using `rspec` for testing and not native testing.
> * [OPTIONAL] Consider using the `timecop` gem to freeze time when running tests. It can help avoid errors: https://github.com/travisjeffery/timecop
> * (frontend) see my comments on SVG and sub-components
> 
> **Hope all is well, and that my comments were helpful. 😸**
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._ _Also, you do not have to implement all requested changes, as this is up to you in the end, but I strongly recommend you implement them._

Thank you for your review @mcrd25  , but in the project requirements there is nothing that says you need tests for the project

**Review by Meltrust — APPROVED** [review]
#### Hi @OnlyUnknown,


# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- SVGs have been cleaned up to other files✔️ 
- Model backend tests have been included✔️ 
- Error handling in controllers has been improved✔️
- Readme has been improved ✔️ 




Cheers and Happy coding!👏👏👏


------

**Comment by ShiroYaksha90** [comment]
Thank you very much @mcrd25 @Meltrust 🙏 Have a pleasant day 😃

**Pull request merged**